### PR TITLE
Harmonise and organise built-in class and module names (for #1114)

### PR DESCRIPTION
**Pull request by TeranIvy**
- Renamed the built-ins definition file `dynamo0p3_builtins_mod.f90` to `lfric_builtins_mod.f90`. I left it in `src/psyclone/parse` for now although I don't think it's the right place for the file as it is LFRic-API-specific. Perhaps it should be moved to `src/psyclone/domain/lfric/parse` or similar (up to the reviewer to decide).
- Added appropriate `use` statements to `lfric_builtins_mod.f90` (carried over from #1108) and various test built-ins definition files in  `src/psyclone/tests/test_files/dynamo0p3`.
- Renamed `dynamo0p3_builtins.py` to `lfric_builtins.py`, harmonised class names from `Dyn*` to `LFRic*` and moved the file to `src/psyclone/domain/lfric`.
- Renamed `dynamo0p3_builtins_test.py` to `lfric_builtins_test.py`, harmonised test names and moved the file to `src/psyclone/tests/domain/lfric`.
- Updated paths and imports in Python modules and the related tests.
- Updated documentation.

### CODE REVIEW

**Comment by TeranIvy** [comment]
@arporter and @rupertford, would it be possible to have this revieweed soonish? I have a few extra built-ins to add for PSyclone 2.0.0 and I need this PR for that work. Thanks!

File: src/psyclone/parse/algorithm.py [hunk 1]
```diff
@@ -449,9 +450,10 @@ def get_builtin_defs(api):
     check_api(api)
 
     if api == "dynamo0.3":
-        from psyclone.dynamo0p3_builtins import BUILTIN_MAP as builtins
-        from psyclone.dynamo0p3_builtins import BUILTIN_DEFINITIONS_FILE as \
-            fname
+        from psyclone.domain.lfric.lfric_builtins import BUILTIN_MAP \
```
Comment (arporter) [inline src/psyclone/parse/algorithm.py:453]: Please could you disable the pylint warning about import-outside-toplevel here.

File: src/psyclone/psyGen.py [hunk 2]
```diff
@@ -2628,8 +2628,17 @@ def zero_reduction_variable(self, parent, position=None):
...
-        '''generate the appropriate code to place after the end parallel
-        region'''
+        '''
+        Generate the appropriate code to place after the end parallel
+        region.
+
+        :param parent: the Node in the AST to which to add new code.
+        :type parent: :py:class:`psyclone.psyir.nodes.Node`
+
+        :raises GenerationError: for an nsupported reduction access in \
```
Comment (arporter) [inline src/psyclone/psyGen.py:2638]: "unsupported"

File: src/psyclone/psyGen.py [hunk 3]
```diff
@@ -2641,8 +2650,8 @@ def reduction_sum_loop(self, parent):
             api_strings = [access.api_specific_name()
                            for access in REDUCTION_OPERATOR_MAPPING]
             raise GenerationError(
-                "unsupported reduction access '{0}' found in DynBuiltin:"
-                "reduction_sum_loop(). Expected one of '{1}'".
+                "Unsupported reduction access '{0}' found in LFRicBuiltIn:"
```
Comment (arporter) [inline src/psyclone/psyGen.py:2653]: Please could you change this to use `six.raise_from()`

File: src/psyclone/tests/domain/lfric/lfric_builtins_test.py [hunk 4]
```diff
@@ -222,20 +224,20 @@ def test_builtin_no_field_args():
 def test_builtin_operator_arg():
     ''' Check that we raise appropriate error if we encounter a built-in
     that takes something other than a field or scalar argument. '''
-    from psyclone.dynamo0p3_builtins import VALID_BUILTIN_ARG_TYPES
-    old_name = dynamo0p3_builtins.BUILTIN_DEFINITIONS_FILE[:]
+    from psyclone.domain.lfric.lfric_builtins import VALID_BUILTIN_ARG_TYPES
```
Comment (arporter) [inline src/psyclone/tests/domain/lfric/lfric_builtins_test.py:227]: Please move this import up.

File: src/psyclone/tests/domain/lfric/lfric_builtins_test.py [hunk 5]
```diff
@@ -271,26 +273,26 @@ def test_builtin_args_not_same_space():
...
 
-def test_dynbuiltincallfactory_str():
-    ''' Check that the str method of DynBuiltInCallFactory works as
+def test_lfricbuiltincallfactory_str():
+    ''' Check that the str method of LFRicBuiltInCallFactory works as
     expected '''
-    from psyclone.dynamo0p3_builtins import DynBuiltInCallFactory
-    dyninf = DynBuiltInCallFactory()
-    assert str(dyninf) == "Factory for a call to a Dynamo built-in."
+    from psyclone.domain.lfric.lfric_builtins import LFRicBuiltInCallFactory
```
Comment (arporter) [inline src/psyclone/tests/domain/lfric/lfric_builtins_test.py:279]: Please mv this import up.

File: src/psyclone/tests/domain/lfric/lfric_builtins_test.py [hunk 6]
```diff
@@ -307,10 +309,10 @@ def test_invalid_builtin_kernel():
             "recognised built-in" in str(excinfo.value))
 
 
-def test_dynbuiltin_str(dist_mem):
+def test_lfricbuiltin_str(dist_mem):
     ''' Check that we raise an error if we attempt to call the __str__
-    method on the parent DynBuiltIn class '''
-    from psyclone.dynamo0p3_builtins import DynBuiltIn
+    method on the parent LFRicBuiltIn class. '''
+    from psyclone.domain.lfric.lfric_builtins import LFRicBuiltIn
```
Comment (arporter) [inline src/psyclone/tests/domain/lfric/lfric_builtins_test.py:315]: Please mv this up.

**Review by arporter — CHANGES_REQUESTED** [review]
Good job Iva.
Just a few minor things to fix.
Tests all OK with compilation. Ditto Examples.
All docs build OK.
I've checked all files with pylint.

File: doc/user_guide/built_ins.rst [hunk 7]
```diff
@@ -273,28 +273,29 @@ Adding new Built-in operations
 ------------------------------
 
  1. Identify the PSyclone source file for the API to be extended. *e.g.* for
-    LFRic (Dynamo0.3) API it is ``src/psyclone/dynamo0p3_builtins.py``.
+    LFRic API it is ``src/psyclone/domain/lfric/lfric_builtins.py``.
```
[the remark this answers is not in the document]
Comment (TeranIvy) [reply doc/user_guide/built_ins.rst:276]: Thanks, corrected.

File: doc/user_guide/built_ins.rst [hunk 8]
```diff
@@ -273,28 +273,29 @@ Adding new Built-in operations
...
  2. Edit this source file to create the class for this new call. It must
     inherit from the API-specific parent class for Built-in operations
-    (``DynBuiltInKern`` for Dynamo0.3).
+    (``LFRicBuiltInKern`` for LFRic API).
```
[the remark this answers is not in the document]
Comment (TeranIvy) [reply doc/user_guide/built_ins.rst:279]: Done.

File: doc/user_guide/built_ins.rst [hunk 9]
```diff
@@ -273,28 +273,29 @@ Adding new Built-in operations
...
     to the ``BUILTIN_MAP`` dictionary in that source file.
  5. Add metadata describing this call to the appropriate file specified in
-    the ``BUILTIN_DEFINITIONS_FILE`` in that source file. For Dynamo0.3
-    this is ``src/psyclone/parse/dynamo0p3_builtins_mod.f90``.
- 6. Add relevant tests to the PSyclone test file for the API to be extended.
-    *e.g.* for Dynamo0.3 it is
-    ``src/psyclone/tests/dynamo0p3_builtins_test.py``. The tests rely on
-    ``single_invoke`` Fortran examples in the relevant
-    ``src/psyclone/tests/test_files/`` subfolder.
+    the ``BUILTIN_DEFINITIONS_FILE`` in that source file. For LFRic API
```
[the remark this answers is not in the document]
Comment (TeranIvy) [reply doc/user_guide/built_ins.rst:284]: We don't have articles in Croatian and Brits do not exactly have any specific rule about when to use them :smile:
Added.

File: doc/user_guide/built_ins.rst [hunk 10]
```diff
@@ -273,28 +273,29 @@ Adding new Built-in operations
...
+    this is ``src/psyclone/parse/lfric_builtins_mod.f90``.
+ 6. Add relevant tests to the PSyclone test files for the API to be extended.
+    *e.g.* for LFRic API they are
```
[the remark this answers is not in the document]
Comment (TeranIvy) [reply doc/user_guide/built_ins.rst:287]: Yep.

File: doc/user_guide/built_ins.rst [hunk 11]
```diff
@@ -273,28 +273,29 @@ Adding new Built-in operations
...
+    * ``src/psyclone/tests/domain/lfric/lfric_builtins_test.py``,
+    * ``src/psyclone/tests/domain/lfric/lfric_integer_builtins_test.py``.
+    The tests rely on ``single_invoke`` Fortran examples in the relevant
+    ``src/psyclone/tests/test_files/`` subdirectory.
  7. Add an appropriate Fortran ``single_invoke`` example for the new
-    Built-in in the relevant ``src/psyclone/tests/test_files/`` subfolder.
-    *e.g.* for Dynamo0.3 it is ``src/psyclone/tests/test_files/dynamo0p3/``.
+    Built-in in the relevant ``src/psyclone/tests/test_files/`` subdirectory.
+    *e.g.* for LFRic API it is ``src/psyclone/tests/test_files/dynamo0p3/``.
```
[the remark this answers is not in the document]
Comment (TeranIvy) [reply doc/user_guide/built_ins.rst:294]: Added.

File: doc/user_guide/built_ins.rst [hunk 12]
```diff
@@ -273,28 +273,29 @@ Adding new Built-in operations
...
     Names of examples follow the template
     ``<category.number>.<subcategory.number>_<built-in_name>.f90``.
-    *e.g.* for Dynamo0.3 API ``<category.number>`` is 15 and
-    ``<built-in_name>`` follows the :ref:`Dynamo0.3 API Built-in naming
+    *e.g.* for LFRic API ``<category.number>`` is 15 and
```
[the remark this answers is not in the document]
Comment (TeranIvy) [reply doc/user_guide/built_ins.rst:297]: Added.

File: doc/user_guide/built_ins.rst [hunk 13]
```diff
@@ -303,7 +304,7 @@ Adding new Built-in operations
 If the API being extended does not currently support any Built-ins
 then the ``BUILTIN_MAP`` and ``BUILTIN_DEFINITIONS_FILE`` module
 variables must be added to the source file for the API.  A Fortran
-module file must be created in the PSyclone src directory (with the
+module file must be created in the PSyclone ``src`` directory (with the
```
[the remark this answers is not in the document]
Comment (TeranIvy) [reply doc/user_guide/built_ins.rst:307]: Agreed. Changed.

File: src/psyclone/tests/domain/lfric/lfric_builtins_test.py (hunk 4 above)
Comment (TeranIvy) [reply src/psyclone/tests/domain/lfric/lfric_builtins_test.py:227]: Moved.

File: src/psyclone/tests/domain/lfric/lfric_builtins_test.py (hunk 5 above)
Comment (TeranIvy) [reply src/psyclone/tests/domain/lfric/lfric_builtins_test.py:279]: Moved.

File: src/psyclone/tests/domain/lfric/lfric_builtins_test.py (hunk 6 above)
Comment (TeranIvy) [reply src/psyclone/tests/domain/lfric/lfric_builtins_test.py:315]: Moved.

File: src/psyclone/dynamo0p3.py [hunk 14]
```diff
@@ -6516,8 +6516,8 @@ def load(self, kern):
         # Loop bounds
         self.set_lower_bound("start")
 
-        from psyclone.dynamo0p3_builtins import DynBuiltIn
-        if isinstance(kern, DynBuiltIn):
+        from psyclone.domain.lfric.lfric_builtins import LFRicBuiltIn
```
[the remark this answers is not in the document]
Comment (TeranIvy) [reply src/psyclone/dynamo0p3.py:6519]: Done.

File: src/psyclone/psyGen.py [hunk 15]
```diff
@@ -2628,8 +2628,17 @@ def zero_reduction_variable(self, parent, position=None):
                                  rhs=zero), position=position)
 
     def reduction_sum_loop(self, parent):
-        '''generate the appropriate code to place after the end parallel
-        region'''
+        '''
+        Generate the appropriate code to place after the end parallel
+        region.
+
+        :param parent: the Node in the AST to which to add new code.
```
[the remark this answers is not in the document]
Comment (TeranIvy) [reply src/psyclone/psyGen.py:2635]: Thanks, corrected.

File: src/psyclone/psyGen.py (hunk 2 above)
Comment (TeranIvy) [reply src/psyclone/psyGen.py:2638]: Oops, corrected.

File: src/psyclone/psyGen.py [hunk 16]
```diff
@@ -2628,8 +2628,17 @@ def zero_reduction_variable(self, parent, position=None):
...
+        :type parent: :py:class:`psyclone.psyir.nodes.Node`
```
[the remark this answers is not in the document]
Comment (TeranIvy) [reply src/psyclone/psyGen.py:2636]: Corrected to `psyclone.f2pygen.SubroutineGen`.

File: src/psyclone/psyGen.py (hunk 3 above)
Comment (TeranIvy) [reply src/psyclone/psyGen.py:2653]: Done.

File: src/psyclone/domain/lfric/lfric_builtins.py [hunk 17]
```diff
@@ -143,20 +143,20 @@ def create(call, parent=None):
         return dofloop
 
 
-class DynBuiltIn(BuiltIn):
+class LFRicBuiltIn(BuiltIn):
```
[the remark this answers is not in the document]
Comment (TeranIvy) [reply src/psyclone/domain/lfric/lfric_builtins.py:146]: Created issue #1140 and added "TODO" here.

**Comment by TeranIvy** [comment]
Re-tested with compilation and checked that the documentation builds. Ready for another review from @arporter.

File: doc/user_guide/built_ins.rst (hunk 9 above)
[the remark this answers is not in the document]
Comment (arporter) [reply doc/user_guide/built_ins.rst]: I'm sure there are rules but I've no idea what they are :-) My son probably knows!

**Review by arporter — APPROVED** [review]
All requested changes have been made.
All tests and examples are OK with compilation.
All docs build cleanly.
I'll merge this once #1122 is on.

**Pull request merged**